api/ocs/__init__.py

```diff
@@ -372,78 +478,248 @@ def parse_nonsidereal_target(data: dict) -> NonsiderealTarget:
+        exposure_times = []
+        coadds = []
+        # wavelengths = []
+
+        instrument = sequence[0][OcsProgramProvider._AtomKeys.INSTRUMENT] # all atoms must have the same instrument
+        offset_lag = OcsProgramProvider._parse_offsets(sequence, 'GPI')
 
-            # Offset information
-            offset_p = OcsProgramProvider._AtomKeys.OFFSET_P
-            offset_q = OcsProgramProvider._AtomKeys.OFFSET_Q
-            p = float(step[offset_p]) if offset_p in step.keys() else None
-            q = float(step[offset_q]) if offset_q in step.keys() else None
+        inst_config = InstrumentConfig([], [], [], [])
+        # Sequence analysis
+        prev = 0
+        n_offsets = 0
+        exptime_groups = False
+        n_pattern = offset_lag
+        for atom_id, step in enumerate(sequence):
+
+            next_atom = False
+            
+            observe_class = step[OcsProgramProvider._AtomKeys.OBS_CLASS]
+            step_time = step[OcsProgramProvider._AtomKeys.TOTAL_TIME] / 1000
+            observed = str_to_bool(step[OcsProgramProvider._AtomKeys.OBSERVED])
 
-            # Any wavelength/filter_name change is a new atom
-            if atom_id == 0 or (atom_id > 0 and
-                                wavelength != float(sequence[atom_id - 1][OcsProgramProvider._AtomKeys.WAVELENGTH])):
+            #Instrument configuration aka Resource
+            # instrument = step[OcsProgramProvider._AtomKeys.INSTRUMENT]
+            fpu, disperser, filter, wavelength  = OcsProgramProvider._parse_instrument_configuration(step, instrument)
+            inst_config.fpu.append(fpu)
+            inst_config.disperser.append(disperser)
+            inst_config.filter.append(filter)
+            inst_config.wavelength.append(wavelength)
+
+            mode = OcsProgramProvider._get_mode(instrument, inst_config)
+            #wavelengths.append(wavelength)
+
+            coadds.append(int(step[OcsProgramProvider._AtomKeys.COADDS]) if OcsProgramProvider._AtomKeys.COADDS in step else 1)
+            exposure_times.append(step[OcsProgramProvider._AtomKeys.EXPOSURE_TIME])
+            
+            # Any wavelength/filter change is a new atom
+            if atom_id == 0 or (atom_id > 0 and inst_config.wavelength[atom_id] != inst_config.wavelength[prev]):
                 next_atom = True
+                logging.info('Atom for wavelength')
 
-            # Patterns:
-            # AB
-            # ABBA
-            if q is not None and n_steps >= 4 and n_steps - atom_id > 3 and n_abba == 0:
-                if (q == float(sequence[atom_id + 3][offset_q]) and
-                        q != float(sequence[atom_id + 1][offset_q]) and
-                        float(sequence[atom_id + 1][offset_q]) == float(sequence[atom_id + 2][offset_q])):
-                    n_abba = 3
+            if step[OcsProgramProvider._AtomKeys.OBSERVE_TYPE].upper() not in ['FLAT', 'ARC', 'DARK', 'BIAS']:
```

Comment:
We have this list of:
```
['FLAT' , 'ARC', 'DARK', 'BIAS']
```
in multiple places. We should move it out of the methods and into the class as a class variable.